### PR TITLE
[FIX] sale_timesheet: delete service tracking for non service products

### PR DESCRIPTION
**Pull request by DjamelTouati**
Steps to reproduce the product:
- Create a new product:
- Select “Service” type
- Go to the Sales tab:
    - Select "Create a task in an existing project" in the `service
    tracking` field
     - the `project` field becomes visible
- Change the Type of the product to “Consumable”
- Try to save

Problem:
The field `project` remains visible and as it is a required field,
we have to fill it in before being able to save the product.
While it is a feature only for service-type products

Solution:
- The `project` field is based on the `service_tracking` field
to become invisible or not. So when the product is not of service type,
we must remove the tracking service so that the project field
becomes invisible.

- Prevent the user from changing the type of products already used in
a sale order

opw-[2858064](https://www.odoo.com/web#id=2858064&view_type=form&model=project.task)

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/sale/models/product_template.py [hunk 1]
```diff
@@ -139,6 +139,11 @@ def _onchange_type(self):
             self.service_type = 'manual'
         return res
 
+    def write(self, values):
+        if 'type' in values and self.sales_count > 0:
```
Comment (Feyensv) [inline addons/sale/models/product_template.py:143]: 1) write is supposed to support multiple records edition `(len(self) > 1)`
`env['product.template'].browse([1,2,3]).write({'type': 'consu'})`
-> ERROR

2) writing the same value shouldn't trigger your warning
```python
product = env['product.product'].create({'type': 'consu'}
env['sale.order.line'].create({'product_id': product.id})
product.write({'type': 'consu'})
```

**Review by Feyensv — COMMENTED** [review]
IMHO, this would be cleaner with a simple warning given by an onchange, instead of an overhead on all product write operations.

File: addons/sale/models/product_template.py (hunk 1 above)
Comment (DjamelTouati) [reply addons/sale/models/product_template.py]: Hello @Feyensv 
Thank you for the review, I agree with points 1 and 2, I fixed them.

However, I don't understand how the onchange solution can work, for example, if we do a RPC call? It is for this reason that I do the verification in the write

Comment (Feyensv) [reply addons/sale/models/product_template.py]: I believe we can consider that modifications through rpc calls (without going through the web interface) are mainly used
by knowledgeable users, and they should be aware that this kind of change is quite risky.

As I see it, this warning is mostly a defensive safeguard against "dumb users making risky changes through the interface" (if the product was used already).   A simple onchange warning should be enough for those users, overrides of core orm methods should be avoided when not really necessary, especially on critical models frequently modified.

Furthermore, the onchange warning has the benefit of being triggered directly when the modification is done in the interface, not only on save (the sooner the warning/error, the less the user will make additional changes on the record).

File: addons/sale/models/product_product.py [hunk 2]
```diff
@@ -34,6 +34,16 @@ def _compute_sales_count(self):
             product.sales_count = float_round(r.get(product.id, 0), precision_rounding=product.uom_id.rounding)
         return r
 
+    @api.onchange('type')
+    def _onchange_type(self):
+        res = super(ProductProduct, self)._onchange_type()
+        if self.sales_count > 0:
```
Comment (Feyensv) [inline addons/sale/models/product_product.py:40]:
```suggestion
        if self.id and self.sales_count > 0:
```

No need to query orders count for non saved product(s) ;).

File: addons/sale/models/product_template.py [hunk 3]
```diff
@@ -137,6 +137,11 @@ def _onchange_type(self):
             if not self.invoice_policy:
                 self.invoice_policy = 'order'
             self.service_type = 'manual'
+        if self.sales_count > 0:
```
Comment (Feyensv) [inline addons/sale/models/product_template.py:140]: same remark :).

File: addons/product/models/product.py [hunk 4]
```diff
@@ -319,6 +319,11 @@ def _onchange_uom(self):
         if self.uom_id and self.uom_po_id and self.uom_id.category_id != self.uom_po_id.category_id:
             self.uom_po_id = self.uom_id
 
+    @api.onchange('type')
+    def _onchange_type(self):
+        # Do nothing but needed for inheritance
```
Comment (Feyensv) [inline addons/product/models/product.py:324]: Needed for inheritance ? `sale_timesheet` depends on sale, why is this needed ?

**Comment by Feyensv** [comment]
Minor change in the warning message, from 'sale order' to 'sales orders' ;).

@robodoo r+

**Comment by DjamelTouati** [comment]
@robodoo r-

**Comment by Feyensv** [comment]
@robodoo delegate+

**Comment by DjamelTouati** [comment]
@robodoo r+ Thanks ;)
